### PR TITLE
fix(components): Update GCP component container to Python 3.7. Fixes #4959 

### LINKED ISSUE #4959: GCP component container is still based on Python 2.7
https://github.com/kubeflow/pipelines/blob/7540ba5c3bf003758a5c2f46ccc586a0367c4dee/components/gcp/container/Dockerfile#L15-L21

GCP component container is still using Python 2.7 which is no longer supported. And the apache-beam SDK is also deprecated by Dataflow: https://cloud.google.com/dataflow/docs/support/sdk-version-support-status#apache-beam-2x-sdks

### PR DESCRIPTION
**Pull request by chensun**
**Description of your changes:**

- Update GCP component container to using Python 3.7, also upgrade apache-beam[gcp] version.
- Fix Dataflow client and sample.

Tested via Dataflow launch_python and launch_template sample notebooks.
Fixes #4959 

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

File: components/gcp/container/Dockerfile [hunk 1]
```diff
@@ -12,19 +12,13 @@
...
-
-RUN pip2 install apache-beam[gcp]==2.10.0
+RUN pip install apache-beam[gcp]
 RUN pip install pandas
 
 ADD build /ml
 WORKDIR /ml
 RUN pip install .
 
-RUN patch /usr/local/lib/python2.7/site-packages/googleapiclient/http.py < /ml/patches/http.patch
```
Comment (numerology) [inline components/gcp/container/Dockerfile:28]: Umm....this is needed for telemetry tracking for KFP on other GCP services. Can we add this back?

**Comment by Ark-kun** [comment]
Thank you for tackling this, Chen.

A general comment: It might be beneficial to split unrelated changes into separate PRs. It looks like this PR changes several things at once - Python 2, APi regionalization and general refactoring.

File: components/gcp/container/component_sdk/python/kfp_component/google/dataflow/_client.py [hunk 2]
```diff
@@ -17,9 +17,9 @@
 
 class DataflowClient:
     def __init__(self):
-        self._df = discovery.build('dataflow', 'v1b3')
+        self._df = discovery.build('dataflow', 'v1b3', cache_discovery=False)
```
[the remark this answers is not in the document]
Comment (chensun) [reply components/gcp/container/component_sdk/python/kfp_component/google/dataflow/_client.py:20]: This seems nontrivial work. To do so I'll have to duplicate a bunch of code inside googleapiclient thats handles different cases for credentials.
Restored the patch change.

File: components/gcp/container/Dockerfile (hunk 1 above)
Comment (chensun) [reply components/gcp/container/Dockerfile:28]: Thanks for the info. I restore thed patch and added some comments.
This isn't ideal as the patch is based on a specific version of the code. It complicates the package version upgrade. But the other option is quite costly. We should probably revisit this later when have bandwidth.

**Comment by chensun** [comment]
> Thank you for tackling this, Chen.
> 
> A general comment: It might be beneficial to split unrelated changes into separate PRs. It looks like this PR changes several things at once - Python 2, API regionalization and general refactoring.

All changes are actually related. As `apache-beam[gcp]` version we pinned previously only works for Python 2.7. So as we upgrade to Python 3.7, we need to use a newer version of `apache-beam[gcp]`. And I found out that new SDK requires a few changes:
- runner needs to be "DatatflowRunner" instead of just "dataflow"
- region is now a required argument, otherwise we got compile time error: 'ValueError: Pipeline has validations errors: Missing required option: region.'

Also the backend now returns job URL in a different format, so we need to change the parsing logic. Though this is not a sdk/container related changes, but it affects e2e runtime results which is currently broken.

**Comment by numerology** [comment]
/lgtm
/approve

Thanks @chensun !

**Pull request merged**